flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java

```diff
@@ -355,20 +356,25 @@ public void close() {
                 // that the stream is closed
                 pos = writeBuffer.length;
 
-                if (outStream != null) {
```

Comment:
The `outStream` and `statePath` should be added the `volatile`.

Async thread created the `outStream`, and then task thread closes the `FsCheckpointStateOutputStream`, and task thread should close the `outStream` and delete the `statePath`. However, the `outStream` and `statePath` didn't add the `volatile`, so task thread may not be able to read `outStream`.